packages/circuits/test/preventDoubleAction.test.ts

```diff
@@ -0,0 +1,348 @@
+    genProposalId,
+} from '@unirep/utils'
+import {
+    Circuit,
+    CircuitConfig,
+    PreventDoubleActionProof,
+    EpochKeyLiteProof,
+} from '../src'
+import { defaultProver } from '../provers/defaultProver'
+
+import {
+    genProofAndVerify,
+    randomData,
+    genPreventDoubleActionCircuitInput,
+} from './utils'
+
+const { STATE_TREE_DEPTH, NUM_EPOCH_KEY_NONCE_PER_EPOCH } =
+    CircuitConfig.default
+
+describe('Prevent double action circuit', function () {
+    this.timeout(300000)
+
+    it('should prove epoch key membership with external nullifier', async () => {
+        for (let nonce = 0; nonce < NUM_EPOCH_KEY_NONCE_PER_EPOCH; nonce++) {
+            const attesterId = BigInt(10210)
+            const epoch = 120958
+            const data = randomData()
+            const id = new ZkIdentity()
+            const tree = new IncrementalMerkleTree(STATE_TREE_DEPTH)
+            const leaf = genStateTreeLeaf(
+                id.secretHash,
+                attesterId,
+                epoch,
+                data
+            )
+            tree.insert(leaf)
+            const externalNullifier = BigInt(1)
+            const circuitInputs = genPreventDoubleActionCircuitInput({
+                id,
+                tree,
+                leafIndex: 0,
+                epoch,
+                nonce,
+                attesterId,
+                data,
+                externalNullifier,
+            })
+            const { isValid, publicSignals, proof } = await genProofAndVerify(
+                Circuit.preventDoubleAction,
+                circuitInputs
+            )
+            expect(isValid).to.be.true
+            expect(publicSignals[0]).to.equal(
+                genEpochKey(id.secretHash, attesterId, epoch, nonce).toString()
+            )
+            expect(publicSignals[1]).to.equal(tree.root.toString())
+            expect(publicSignals[2]).to.equal(
+                EpochKeyLiteProof.buildControl({
+                    attesterId,
+                    epoch,
+                    nonce,
+                }).toString()
+            )
+            const nullifier = genNullifier(
+                id.identityNullifier,
+                externalNullifier
+            )
+            expect(publicSignals[3]).to.equal(nullifier.toString())
+            const identityCommitment = genProposalId(nullifier, id.trapdoor)
+            expect(publicSignals[4]).to.equal(identityCommitment.toString())
+
+            const p = new PreventDoubleActionProof(publicSignals, proof)
+            expect(p.epoch.toString()).to.equal(epoch.toString())
+            expect(p.nonce.toString()).to.equal('0')
+            expect(p.revealNonce.toString()).to.equal('0')
+            expect(p.attesterId.toString()).to.equal(attesterId.toString())
+        }
+    })
+
+    // it('should reveal nonce value', async () => {
+    //     for (let nonce = 0; nonce < NUM_EPOCH_KEY_NONCE_PER_EPOCH; nonce++) {
+    //         const attesterId = BigInt(10210)
```

Comment:
Also make sure to add tests for `PreventDoubleActionProof`. Like pass in a unique value for each input and check that the variables in the `PreventDoubleActionProof` match what is expected.

Reply:
It needs a bit more, it should check all the output variables, like `epochKey` and `identityCommitment` as well as handle the case of `revealNonce` being both `0` and `1`.

We should also consider not forcing the user to reveal an epoch key, but that can be added in later.

Reply:
so does this look right for checking `epochKey` field?
`expect(p.epochKey.toString()).to.equal(genEpochKey(id.secretHash, attesterId, epoch, nonce).toString())`
Also I looked for a reference, but I didn't find one. Why do we not check public output fields of a proof in other test?

Reply:
Yes that looks correct.

Ah yes in other tests we seem to only check the control values. That's because someone (me) was not paying attention. Best to check the output mappings in at least 1 place to avoid silently breaking it by rearranging outputs.